docs: add Interactive Apps guide (#94020)
vercel/next.js: docs/01-app/02-guides/interactive-apps.mdx

- [`useTransition`](https://react.dev/reference/react/useTransition) and [`startTransition`](https://react.dev/reference/react/startTransition) mark async updates as transitions and expose `isPending`. `useTransition` also catches errors thrown inside the async callback and bubbles them to the nearest error boundary.
- [`useActionState`](https://react.dev/reference/react/useActionState) wraps a form action with pending state, reset behavior, and result handling.
- [`'use cache'`](/docs/app/api-reference/directives/use-cache) with [`cacheTag`](/docs/app/api-reference/functions/cacheTag) and [`updateTag`](/docs/app/api-reference/functions/updateTag) keeps reusable reads warm across navigations, and [runtime prefetching](/docs/app/guides/prefetching) fills the next destination before the click.

Most patterns mix two or more of these. Reach for whichever primitive fits the constraint you're solving.

Related guides:

- [Streaming](/docs/app/guides/streaming) for loading boundaries and `<Suspense>` patterns
- [Instant Navigation](/docs/app/guides/instant-navigation) for validating that navigations stay instant

> - [View Transitions](/docs/app/guides/view-transitions) for animating state changes

Reviewer: Should we add a callout/x-ref about INP? the contents of this guide are the kind of thing that can also help people with their INP metrics, although often this can happen from heavy client modules, or blocking roundtrips - could be good to do a quick research and add a note about it

Author: Updated, check last two commits!

---
title: Building interactive apps
description: Learn how to build responsive interactions with Server Functions, transitions, optimistic UI, action props, and pending feedback.
nav_title: Interactive apps
related:
  description: Learn more about the APIs used in this guide.
  links:
    - app/api-reference/functions/refresh
    - app/api-reference/directives/use-cache
    - app/api-reference/functions/cacheTag
    - app/api-reference/functions/updateTag
    - app/guides/streaming
---

> Server Components render the current server data. Client Components start interactions before the next server render is ready. Without additional state, the UI keeps showing the old server props while a Server Function, form action, or navigation is still running.

Reviewer: Wonder if this intro fits a bit better.

```suggestion
When a user interaction requires server-side work, the result is not available immediately. Network requests take an unknown amount of time to complete and may succeed or fail. While that work is in progress, the client should provide feedback and update the UI accordingly. Without it, users are left looking at stale data and may wonder whether their action had any effect.
```

Reviewer: Or maybe this 🤔 - up to you hehe

```suggestion
When a user interaction requires server-side work, the result is not available immediately. Network requests take an unknown amount of time to complete and may succeed or fail. While waiting, the client should provide feedback and respond to those states. Otherwise, users are left looking at stale data and may wonder whether anything is happening.
```

If you navigate to a task now, the page shell paints immediately, the task header streams in when `getTask` resolves, and the comments section streams in after that when `getComments` resolves.

For more on streaming patterns, see the [Streaming](/docs/app/guides/streaming) guide.

### Step 2: Respond instantly to a toggle

Each task card has a priority indicator. Clicking it cycles through low, medium, and high.

The button renders the `priority` prop from the server. Calling a Server Function starts the mutation, but it does not change that prop in the current render.

Here the Server Function is called directly:

```tsx filename="components/task-card.tsx"
'use client'

import { cyclePriority } from '@/data/actions/task'

export function TaskCard({ id, priority }) {
  return (
    <button onClick={() => cyclePriority(id)} className={priorityDot[priority]}>
      {priority}
    </button>
  )
}
```

If you click the dot, the Server Function runs, but the button keeps rendering the old `priority` prop until the next server render arrives.

> Replace the direct call with `useOptimistic` and `startTransition`. `useOptimistic` renders a temporary value while the Server Function is pending. `startTransition` scopes the async work so React knows when to keep or discard the optimistic value:

Reviewer: A little nit, trying to replace `scopes the async work` 

```suggestion
Replace the direct call with `useOptimistic` and `startTransition`. `useOptimistic` provides a value to render in place of the stale prop. `startTransition` runs Server Function as part of a transition. The optimistic value applies for as long as that transition is pending.
```

Something like that, could be a little easier to read, without having to learn what startTransition does

```tsx filename="components/task-card.tsx"
'use client'

import { startTransition, useOptimistic } from 'react'
import { cyclePriority } from '@/data/actions/task'
import { PRIORITY_CYCLE } from '@/lib/data'

export function TaskCard({ id, priority }) {
  const [optimisticPriority, setOptimisticPriority] = useOptimistic(priority)

  function handlePriority() {
    startTransition(async () => {
      setOptimisticPriority(PRIORITY_CYCLE[optimisticPriority])
      await cyclePriority(id)
    })
  }

  return (
    <button
      onClick={handlePriority}
      className={priorityDot[optimisticPriority]}
    >
      {optimisticPriority}
    </button>
  )
}
```

The `setOptimisticPriority` call updates the UI on the current frame with the next value in the cycle. Reading from `optimisticPriority` (not the prop) means rapid double-clicks cycle correctly instead of reading a stale closure value.

> `startTransition` wraps the async work so React tracks it as a pending transition. If the Server Function throws, the optimistic value reverts. The standalone `startTransition` import does not bubble errors to error boundaries. If you need errors to reach an error boundary, use `useTransition` instead (shown in Step 3).

Reviewer: Paired with the comment above, because the first sentence here felt repetitive-ish

```suggestion
If the Server Function inside `startTransition` throws, the transition ends and the optimistic value reverts. The standalone `startTransition` import does not bubble errors to error boundaries. If you need errors to reach an error boundary, use `useTransition` instead (shown in Step 3).
```

  { label: 'Frontend', value: 'frontend' },
  { label: 'Backend', value: 'backend' },
]

export function LabelFilter() {
  const router = useRouter()
  const searchParams = useSearchParams()
  const [isPending, startTransition] = useTransition()
  const current = searchParams.get('label') ?? null

  function filterAction(value: string | null) {
    const params = new URLSearchParams(searchParams.toString())
    if (value) {
      params.set('label', value)
    } else {
      params.delete('label')
    }
    startTransition(() => {
      router.push(`/?${params.toString()}`)
    })
  }

  return (
    <div data-pending={isPending ? '' : undefined}>
      <ChipGroup items={labels} value={current} changeAction={filterAction} />
    </div>
  )
}
```

The `data-pending` attribute on the wrapper lets parent components react via CSS. The board on the home page uses `group-has-data-pending:opacity-50` to dim while the transition runs, without replacing the board with a skeleton.

If you click a chip now, the board dims while the filtered data loads. The `data-pending` attribute stays on the wrapper until the navigation completes, then the board updates with the filtered tasks.

> But the chip itself still doesn't highlight until after the transition. The `isPending` boolean tells us the navigation is in flight, but it doesn't help the chip show a selected state before the new data arrives.

Reviewer: ```suggestion
But the chip itself still doesn't highlight until the navigation completes. The `isPending` boolean tells us the navigation is in flight, but it doesn't help the chip show a selected state before the new data arrives.
````

As task cards scroll into view, the detail page prefetches into the Client Cache. By the time the user clicks, the destination paints instantly. If a card hasn't been prefetched yet, the existing `<Suspense>` fallbacks still cover the gap.

#### Choosing what to cache and prefetch

Three layers combine for instant repeat navigation, each with a different cost:

| Layer                                 | Cost                         | Apply to                                                 |
| ------------------------------------- | ---------------------------- | -------------------------------------------------------- |
| `<Suspense>`                          | None                         | Any uncached read that can stream                        |
| `'use cache'` + `cacheTag`            | Server memory                | Reads that return the same value for the next viewer     |
| `unstable_prefetch = 'force-runtime'` | Server work per visible link | Destinations where the read is keyed on the real request |

## Next steps

Each pattern adds state for a different kind of pending work:

> | Pattern            | Use it when                                                   |

Reviewer: I think we could switch to a situation -> API look up table?

```suggestion
| Situation | Use |
| --- | --- |
| One value should update before the server responds | `useOptimistic` |
| A URL change triggers a new server render | `useTransition` |
| Slow data should stream in without blocking the page shell | `<Suspense>` |
| A reusable component owns part of the pending UI | Action prop |
| A new item should appear before it exists on the server | `useOptimistic` (pending-only list) |
| A collection needs a temporary structural change | `useOptimistic` (reducer form) |
| A form needs pending state, reset behavior, and result state | `useActionState` |
| A parent should style itself while a child action is pending | `data-pending` attribute |
| A reusable read should survive across renders and navigations | `'use cache'` + `cacheTag` / `updateTag` |
| A cached destination should be filled before the user clicks | Runtime prefetch (`unstable_prefetch = 'force-runtime'`) |
```
